File: bindings/java/src/test/java/com/truespar/siftx/TestPaths.java
package com.truespar.siftx;

import java.io.File;
import java.nio.file.*;

final class TestPaths {
    private TestPaths() {}

    private static final String REPO_ROOT = findRepoRoot();

    static final String EXIFTOOL_IMAGES = REPO_ROOT + "/testdata/exiftool-images";
    static final String EXIF_SAMPLES = REPO_ROOT + "/testdata/exif-samples";
    static final String POPPLER_TEST = REPO_ROOT + "/testdata/poppler-test";

    static boolean hasExifToolImages() { return Files.isDirectory(Path.of(EXIFTOOL_IMAGES)); }
    static boolean hasExifSamples() { return Files.isDirectory(Path.of(EXIF_SAMPLES)); }
    static boolean hasPopplerTest() { return Files.isDirectory(Path.of(POPPLER_TEST)); }

    private static String findRepoRoot() {
        var dir = Path.of("").toAbsolutePath();
        while (dir != null) {
            if (Files.exists(dir.resolve("Cargo.toml"))) {
                return dir.toString();
            }
            dir = dir.getParent();
        }
        // Fallback
        return Path.of("").toAbsolutePath().resolve("../..").normalize().toString();
    }
}

File: bindings/csharp/tests/SiftX.Benchmarks/TestPaths.cs
namespace SiftX.Benchmarks;

internal static class TestPaths
{
    private static readonly string RepoRoot = FindRepoRoot();

    public static string ExifToolImages => Path.Combine(RepoRoot, "testdata", "exiftool-images");
    public static string ExifSamples => Path.Combine(RepoRoot, "testdata", "exif-samples");
    public static string PopplerTest => Path.Combine(RepoRoot, "testdata", "poppler-test");

    public static bool HasExifToolImages => Directory.Exists(ExifToolImages);
    public static bool HasExifSamples => Directory.Exists(ExifSamples);
    public static bool HasPopplerTest => Directory.Exists(PopplerTest);

    private static string FindRepoRoot()
    {
        var dir = Directory.GetCurrentDirectory();
        while (dir != null)
        {
            if (File.Exists(Path.Combine(dir, "Cargo.toml")))
                return dir;
            dir = Directory.GetParent(dir)?.FullName;
        }
        return Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", "..", "..", "..", "..", ".."));
    }
}
